extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/devmode/QuteErrorPageSetup.java

```diff
@@ -0,0 +1,223 @@
+        int lineNumber = getLine(origin);
+        int lineCharacterStart = getLineCharacterStart(origin);
+
+        List<String> sourceLines = new ArrayList<>();
+        try (BufferedReader in = getBufferedReader(templateId)) {
+            String line = null;
+            while ((line = in.readLine()) != null) {
+                sourceLines.add(escaper.escape(line).replace(" ", "&nbsp;"));
+            }
+        } catch (Exception e) {
+            LOG.warn("Unable to read the template source: " + templateId, e);
+        }
+
+        List<Integer> realLines = new ArrayList<>();
+        boolean endLinesSkipped = false;
+        if (sourceLines.size() > 15) {
+            // Line with error plus few surrounding lines
+            int fromIndex = lineNumber > 7 ? (lineNumber - 8) : 0;
+            int toIndex = (lineNumber + 7) > sourceLines.size() ? sourceLines.size() : lineNumber + 7;
+            for (int j = fromIndex; j < toIndex; j++) {
+                // e.g. [10,11,12]
+                realLines.add(j + 1);
+            }
+            sourceLines = sourceLines.subList(fromIndex, toIndex);
+            endLinesSkipped = toIndex != sourceLines.size();
+        } else {
+            for (int j = 0; j < sourceLines.size(); j++) {
+                // [1,2,3]
+                realLines.add(j);
+            }
+        }
+
+        return problemTemplate
+                .data("problemIndex", index)
+                .data("title", messageLines[0])
+                .data("description", Arrays.stream(messageLines).skip(1).collect(Collectors.joining("<br>")))
+                .data("sourceLines", sourceLines)
+                .data("lineNumber", lineNumber)
+                .data("lineCharacterStart", lineCharacterStart)
+                .data("realLines", realLines)
+                .data("endLinesSkipped", endLinesSkipped)
+                .data("space", " ")
+                .render();
+    }
+
+    static String htmlPadRight(String s, int n) {
+        return String.format("%-" + n + "s", s).replace(" ", "&nbsp;");
+    }
+
+    private BufferedReader getBufferedReader(String templateId) throws IOException {
+        for (Path resource : hotReplacementContext.getResourcesDir()) {
+            // src/main/resources/templates
+            Path templates = resource.resolve("templates");
+            if (Files.exists(templates)) {
+                // src/main/resources/templates/items.html
+                Path template = templates.resolve(templateId.replace("\\", "/"));
+                if (Files.exists(template)) {
+                    return Files.newBufferedReader(template);
+                }
+            }
+        }
+        throw new IllegalStateException("Template source not available");
+    }
+
+    private Object getOrigin(Throwable t) {
+        Object origin = null;
+        try {
+            Method getOrigin = t.getClass().getClassLoader().loadClass(TEMPLATE_EXCEPTION).getMethod("getOrigin");
```

Comment:
I think @stuartwdouglas should look at this and the following methods and tell us if this is normal, or a bug.

Reply:
The templates are built at deployment time, the error page handler runs at runtime, which means we are in a different CL to the one that built the template.

Reply:
So it's fine then, OK.